intellij-plugin-structure/structure-teamcity-actions/src/main/kotlin/com/jetbrains/plugin/structure/teamcity/action/TeamCityActionPluginManager.kt

```diff
@@ -0,0 +1,199 @@
+    // TODO: Probably the file name should not be hardcoded, see related comments with questions below
+    const val ACTION_DESCRIPTOR_FILE = "action.yml"
+
+    private val LOG: Logger = LoggerFactory.getLogger(TeamCityActionPluginManager::class.java)
+
+    fun createManager(
+      extractDirectory: Path = Paths.get(Settings.EXTRACT_DIRECTORY.get()),
+    ): TeamCityActionPluginManager {
+      extractDirectory.createDir()
+      return TeamCityActionPluginManager(extractDirectory)
+    }
+  }
+
+  // TODO: What should we expect as the input 'pluginFile' here?
+  // 1. A single YAML file that contains the action description
+  // 2. An archive that contains the YAML file (and maybe something else, e.g. the icon). What should
+  //    be the name of the YAML file in this case?
+  // 3. Maybe something else?
+  override fun createPlugin(pluginFile: Path): PluginCreationResult<TeamCityActionPlugin> {
+    require(pluginFile.exists()) { "TeamCity Action plugin file $pluginFile does not exist" }
+    return when {
+      pluginFile.isZip() -> loadDescriptorFromZip(pluginFile)
+      else -> PluginCreationFail(IncorrectPluginFile(pluginFile.simpleName, ".zip archive"))
+    }
+  }
+
+  private fun loadDescriptorFromZip(pluginFile: Path): PluginCreationResult<TeamCityActionPlugin> {
+    val sizeLimit = Settings.TEAMCITY_ACTION_PLUGIN_SIZE_LIMIT.getAsLong()
+    if (Files.size(pluginFile) > sizeLimit) {
+      return PluginCreationFail(FileTooBig(pluginFile.simpleName, sizeLimit))
+    }
+
+    val tempDirectory = Files.createTempDirectory(extractDirectory, "teamcity_action_")
+    return try {
+      extractZip(pluginFile, tempDirectory, sizeLimit)
+      loadPluginInfoFromDirectory(tempDirectory)
+    } catch (e: DecompressorSizeLimitExceededException) {
+      return PluginCreationFail(PluginFileSizeIsTooLarge(e.sizeLimit))
+    } finally {
+      tempDirectory.deleteLogged()
+    }
+  }
+
+  private fun loadPluginInfoFromDirectory(pluginDirectory: Path): PluginCreationResult<TeamCityActionPlugin> {
+    val descriptorFile = pluginDirectory.resolve(ACTION_DESCRIPTOR_FILE)
+    if (!descriptorFile.exists()) {
+      return PluginCreationFail(PluginDescriptorIsNotFound(ACTION_DESCRIPTOR_FILE))
+    }
+    val descriptorContent = descriptorFile.readText()
+    val mapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+    val descriptor: TeamCityActionDescriptor
+    try {
+      descriptor = mapper.readValue(descriptorContent, TeamCityActionDescriptor::class.java)
+    } catch (e: Exception) {
+      LOG.warn("Failed to parse TeamCity Action. Error: ${e.message}", e)
+      return PluginCreationFail(ParseYamlProblem)
+    }
+    return createPlugin(descriptor)
+  }
+
+  private fun createPlugin(descriptor: TeamCityActionDescriptor): PluginCreationResult<TeamCityActionPlugin> {
+    try {
+      val validationResult = validateTeamCityAction(descriptor)
+      if (validationResult.any { it.level == PluginProblem.Level.ERROR }) {
```

Comment:
Prehaps use the [extension function](https://github.com/JetBrains/intellij-plugin-verifier/blob/1d78511ebbb691f88381a5a47e236015a5fe43ff/intellij-plugin-structure/structure-base/src/main/kotlin/com/jetbrains/plugin/structure/base/problems/PluginProblems.kt#L33)

Reply:
Done, thank you